src/package/providers/nix.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright 2015-2017 Intecture Developers. See the COPYRIGHT file at the
+// top-level directory of this distribution and at
+// https://intecture.io/COPYRIGHT.
+//
+// Licensed under the Mozilla Public License 2.0 <LICENSE or
+// https://www.tldrlegal.com/l/mpl-2.0>. This file may not be copied,
+// modified, or distributed except according to those terms.
+
+//! Nix package provider
+
+use command::{Command, CommandResult};
+use error::Result;
+use host::Host;
+use super::*;
+
+pub struct Nix;
+
+impl Provider for Nix {
+    fn get_providers(&self) -> Providers {
+        Providers::Nix
+    }
+
+    fn is_active(&self, host: &mut Host) -> Result<bool> {
+        let cmd = Command::new("which nix-env");
+        let result = try!(cmd.exec(host));
+
+        Ok(result.exit_code == 0)
+    }
+
+    fn is_installed(&self, _: &mut Host, _: &str) -> Result<bool> {
+        // Nix packages can be installed by name, attribute path
+        // or full Nix store path. They are not all queried the same
+        // way. Also querying by name is very slow. Considering that
+        // the Nix package manager is atomic and that it doesn't
+        // reinstall a package that's already on the system, it's
+        // better to just assume that the package is not installed
+        // and try to install it when needed.
+        Ok(false)
```

Comment:
Hang on, I have just thought of a way to do this in a reliable but also efficient way. Using `nix-env --install --dry-run`. I'm ashamed I didn't think of dry running before. I have been focusing on `nix-env --query` aka `nix-env -q` which really doesn't work for our use case.